### DISCUSSION TITLE
Debugging mSBC no audio

### DISCUSSION THREAD

**Discussion by mrcomcast123** [Arkq/bluez-alsa#520] [General]
Hello, I need some advice on how to debug no audio when playing using SCO (mSBC).

My dev environment: Ubuntu 20.04.3 LTS, kernel 5.11.0-43-generic, bluez 5.56, bluez-alsa master @ f055cf18157325629d60186693e36d350e8d5b71, Morpheus 260 HP4500 headset

SCO (CVSD) and A2DP (SBC) playback work fine, but after configuring with --enable-msbc, SCO (mSBC) doesn't produce sound nor any errors that i can see.  I have --enable-debug and I see msbc is being activated and confirmed the msbc decode path is taken.  There are also no obvious errors from aplay or bluealsa or dmesg that I see.  I have confirmed using pulseaudio and pipewire that my headset can playback mSBC.  

Here are my commands:
```
sudo src/bluealsa -p a2dp-source -p hfp-ag

bluealsa-aplay --list-pcms
bluealsa:SRV=org.bluealsa,DEV=E7:D0:09:A4:FE:7D,PROFILE=sco
    HP4500, trusted audio-card, capture
    SCO (mSBC): S16_LE 1 channel 16000 Hz
bluealsa:SRV=org.bluealsa,DEV=E7:D0:09:A4:FE:7D,PROFILE=a2dp
    HP4500, trusted audio-card, playback
    A2DP (SBC): S16_LE 2 channels 48000 Hz
bluealsa:SRV=org.bluealsa,DEV=E7:D0:09:A4:FE:7D,PROFILE=sco
    HP4500, trusted audio-card, playback
    SCO (mSBC): S16_LE 1 channel 16000 Hz
    
aplay -D bluealsa:DEV=E7:D0:09:A4:FE:7D,PROFILE=sco ~/Downloads/sound-effects-library-wow-sound-effect-anime-style.wav
```

I'm attaching logs of working (cvsd) and non-working(msbc) cases if anybody has a spare moment to look.  

[bluezalsa-cvsd-working-log.txt](https://github.com/Arkq/bluez-alsa/files/7817479/bluezalsa-cvsd-working-log.txt)
[bluezalsa-mSBC-not-working-log.txt](https://github.com/Arkq/bluez-alsa/files/7817480/bluezalsa-mSBC-not-working-log.txt)

Any guidance on how to debug this further would be much appreciated.

Thank You

**Comment by arkq** [comment]
>  I have confirmed using pulseaudio and pipewire that my headset can playback mSBC.

Have you confirmed that in the same environment (the same board/PC, the same BT chip)?

The first thing I would check is the audio capture using mSBC audio codec. You can use arecord with the same -D option as with aplay. Without any redirection to a file, you will see in the console that the output will change when poking into headset microphone. If there is no incoming mSBC audio, then much deeper debugging will be required.

EDIT:
It is quit likely that in order to get mSBC working you will need to upgrade kernel version. Right now you are using 5.11.0 but required patch (for mSBC over USB) needs >= 5.11.1 (see this: https://github.com/Arkq/bluez-alsa/issues/400#issuecomment-784394343).

**Reply by mrcomcast123** [reply]
arecord did show mic output.  I did some more tinkering and switching back and forth between pulse/pipewire/bluealsa, and I don't know what actually changed, but finally bluealsa aplay to SOC(mSBC) started working.  Now I'm trying to figure out what I accidently fixed to make it work.  Thanks
